### PR TITLE
Initial commit of speech detection

### PR DESCRIPTION
**Pull request by nguyer**
I think these are sensible default parameters, but we may want to open up more of the `AudioLevelChangeOptions` to end users of the SDK if they have a specific need to tune it.

### CODE REVIEW

File: src/audioLevelDetector.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      this.timeThreshold = config.timeThreshold;
+    }
+    if (config.maxEmitInterval) {
+      this.maxEmitInterval = config.maxEmitInterval;
+    }
+    if (config.sampleInterval) {
+      this.sampleInterval = config.sampleInterval;
+    }
+
+    // Chrome hack
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:42]: how does this work on other browsers that already work(-ish)?

**Comment by michaelcaldwell** [comment]
jest tests?

File: src/audioLevelDetector.ts (hunk 1 above)
Comment (nguyer) [reply src/audioLevelDetector.ts:42]: It has no effect on other browsers (a.k.a Firefox)

**Comment by nguyer** [comment]
@michaelcaldwell Good point about tests... Off the top of my head, I'm not sure how to go about writing unit tests for this. Do you have any suggestions on an approach there?

**Comment by michaelcaldwell** [comment]
@nguyer I would inject the dependencies into the constructor, vs. creating them inside the constructor, or otherwise, create simple factory methods for those which you can then mock.  In general, call some function (functional unit, unit under test), making the code more functional as required to be well tested, then check properties/state after calling the function.  Just prove that each function behaves as expected.  If we ever change behavior or change properties in the future, the test should then catch it, i.e. break.

**Comment by nguyer** [comment]
@michaelcaldwell Overall test coverage has increased from 10% to 24% now with the coverage provided on this new class. The new class has 80% coverage.

File: src/audioLevelDetector.test.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import AudioLevelDetector from "./audioLevelDetector";
+import { sleep } from "./utils";
+
+type fakeMediaStream = MediaStream;
+
+class MockAudioContext {
+  createMediaStreamSource() {
+    return { connect: () => {} };
+  }
+  createAnalyser() {
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.test.ts:11]: "Analyzer"?  you use that spelling elsewhere.

File: src/audioLevelDetector.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+export default class AudioLevelDetector extends EventEmitter {
+  private start: number = 0;
+  private lastEmitTime: number = 0;
+  private timeThreshold = 500;
+  private silenceAmplitudeThreshold = 0.2;
+  private highAmplitudeThreshold = 0.5;
+  private sampleInterval: number = 100; // ms
+  private maxEmitInterval: number = 500; // ms;
+  private analyzerNode: AnalyserNode;
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:20]: using both the "analyze" and "analyse" root spelling in multiple places, we should really pick one spelling or the other

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export const sleep = (ms: number) => {
```
Comment (michaelcaldwell) [inline src/utils.ts:1]: there isn't a sleep(millis) function implemented already in javascript or some third party lib that we could use (and maybe are already loading)?

File: src/audioLevelDetector.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  sampleInterval?: number;
+  maxEmitInterval?: number;
+}
+
+export default class AudioLevelDetector extends EventEmitter {
+  private start: number = 0;
+  private lastEmitTime: number = 0;
+  private timeThreshold = 500;
+  private silenceAmplitudeThreshold = 0.2;
+  private highAmplitudeThreshold = 0.5;
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:17]: There is no optional property in config for overriding this, but there is for the other four fields here.

File: src/audioLevelDetector.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  private currentAudioLevel: AudioLevel = AudioLevel.SILENT;
+  private previousAudioLevel: AudioLevel | undefined;
+
+  constructor(config: AudioLevelDetectorOptions) {
+    super();
+
+    const context = new AudioContext();
+
+    if (config.amplitudeThreshold) {
+      this.silenceAmplitudeThreshold = config.amplitudeThreshold;
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:30]: config prop should be named same as the field, like the others

File: src/audioLevelDetector.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  private maxEmitInterval: number = 500; // ms;
+  private analyzerNode: AnalyserNode;
+  private currentAudioLevel: AudioLevel = AudioLevel.SILENT;
+  private previousAudioLevel: AudioLevel | undefined;
+
+  constructor(config: AudioLevelDetectorOptions) {
+    super();
+
+    const context = new AudioContext();
+
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:28]: would be a little better if the config fields were checked and assigned to fields in the same order that the fields are declared in above, or alternatively, you could order the fields alphabetically.  It's a little hard as-is to map config field logic to fields, and grok it.

File: src/audioLevelDetector.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    // Chrome hack
+    new Audio().srcObject = config.mediaStream;
+    const source = context.createMediaStreamSource(config.mediaStream);
+    const analyser = context.createAnalyser();
+    analyser.fftSize = 2048;
+    analyser.minDecibels = -90;
+    analyser.maxDecibels = -10;
+    analyser.smoothingTimeConstant = 0.85;
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:49]: should these values be optionally configurable as well?

File: src/audioLevelDetector.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    // Iterate over each of the samples
+    for (let i = 0; i < bufferLength; i++) {
+      const sampleValue = this.normalizeSample(dataArray[i]);
+      this.analyzeSample(sampleValue);
+      this.emitCurrentAudioLevel();
+    }
+  }
+
+  normalizeSample(sample: number) {
+    return sample / 128;
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.ts:69]: why 128?

File: src/audioLevelDetector.test.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  };
+});
+
+afterAll(() => {
+  //@ts-ignore
+  global.AudioContext = realAudioContext;
+  global.setInterval = realInterval;
+});
+
+test("test emit silent at start", () => {
```
Comment (michaelcaldwell) [inline src/audioLevelDetector.test.ts:35]: nice test suite

File: src/audioLevelDetector.test.ts (hunk 2 above)
Comment (nguyer) [reply src/audioLevelDetector.test.ts:11]: I agree. Unfortunately it is spelled "wrong" in the Web Audio API itself... I suppose we could also spell it "wrong", just to be consistent 😐

File: src/audioLevelDetector.ts (hunk 3 above)
Comment (nguyer) [reply src/audioLevelDetector.ts:20]: I agree. Unfortunately it is spelled "wrong" in the Web Audio API itself... I suppose we could also spell it "wrong", just to be consistent 😐

File: src/utils.ts (hunk 4 above)
Comment (nguyer) [reply src/utils.ts:1]: There is not an "awaitable" sleep built into JavaScript. There probably is one in a library somewhere, but it is trivial to implement, and I'd say not worth the tradeoff to take an additional dependency.

File: src/audioLevelDetector.ts (hunk 5 above)
Comment (nguyer) [reply src/audioLevelDetector.ts:17]: Good point. I will add this.

File: src/audioLevelDetector.ts (hunk 8 above)
Comment (nguyer) [reply src/audioLevelDetector.ts:49]: Possibly. I'm not sure what "safe" ranges are for them though. I suggest leaving them as is for now, and making them configurable later if necessary. I think it's less risky than leaving them wide open and allowing someone to set an inappropriate value.

File: src/audioLevelDetector.ts (hunk 3 above)
Comment (michaelcaldwell) [reply src/audioLevelDetector.ts:20]: I would spell it correctly everywhere you can, i.e. our vars that are assigned from their lib or are an instance of their object, then maybe just add a comment that points out their odd spelling.

File: src/utils.ts (hunk 4 above)
Comment (michaelcaldwell) [reply src/utils.ts:1]: true.  sdk dependencies are _de minimis_, just the lang stuff and packing and testing stuff we need, and the rtc stuff we need, so we should strive to keep it that way.  so this is good then.  But could we put it in a time.ts file?  I dislike the whole utils/utilities anti-pattern, which of course is commonplace, and commonly wrong, imo.  It just means the developer(s) haven't thought through, or thought enough, their libs design.

File: src/audioLevelDetector.ts (hunk 8 above)
Comment (michaelcaldwell) [reply src/audioLevelDetector.ts:49]: I would elaborate the "// Chrome hack" comment then, something like:  "// This block of code is necessary for audio level detection to work in Chrome.  We recommend leaving these values as-is."

File: src/audioLevelDetector.ts (hunk 3 above)
Comment (nguyer) [reply src/audioLevelDetector.ts:20]: I think I'll just spell it the way the API spells it just to be consistent. Even if it's un-American...

**Pull request merged**